### ISSUE TITLE
setting boxes=False in segmentation doesn't help

### ISSUE THREAD

**Issue by UTKARSHS1994** [ultralytics/ultralytics#4874]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,

I am working on a segmentation problem. I would like to get inference on the images using trained model with masks but without bounding box. In the official documentation, there is one argument exactly for my need:  

![image](https://github.com/ultralytics/ultralytics/assets/54258787/6f54de9d-3afa-4d71-938e-708666354c72)

However, when I try to get inference, setting boxes=False, I still get the bounding box while I am looking only for mask without bounding box. 
What is the purpose of this argument if it cannot be used for hiding boxes? Or am I missing something?

Link to the prediction arguments official documentation- https://docs.ultralytics.com/modes/predict/#inference-arguments

Regards,
Utkarsh

**Comment by RizwanMunawar** [comment]
@UTKARSHS1994 I am unable to replicate the problem you're experiencing. I used the command mentioned for the segmentation with `boxes=False`.
```
yolo task=segment mode=predict source="ultralytics/videos/segment_test.mp4" boxes=False show=True
```

### Output
![segmentation_test](https://github.com/ultralytics/ultralytics/assets/62513924/6793af30-3611-41a7-9c86-69f6e8cc9083)

**Comment by UTKARSHS1994** [comment]
Hi,
I have a segmentation model which I trained using the YOLO architecture by running it on custom dataset of underwater images for 100 epochs.

## Model training
model = YOLO('yolov8m-seg.yaml').load('yolov8m-seg.pt') 
results=model.train(data=config_path, epochs=100)

## Code for inference
import os
import matplotlib.pyplot as plt
from PIL import Image
!pip install ultralytics==8.0.157
from ultralytics import YOLO

img=plt.imread('/path/to/my/underwater_image.jpg')
plt.imshow(img)
![image](https://github.com/ultralytics/ultralytics/assets/54258787/bd041007-3bbb-43e0-9cd3-1b09d39edb23)

model=YOLO('/path/to/model_weights/best.pt')
results=model('/path/to/my/underwater_image.jpg', boxes=False)
image=Image.fromarray(results[0].plot()[...,::-1])
plt.imshow(image)
![image](https://github.com/ultralytics/ultralytics/assets/54258787/3e97379e-d3e3-4cb2-b6f4-d0936c482683)

Hope this is helpful. How do I only have masks and no boxes?

Regards,
Utkarsh

**Comment by glenn-jocher** [comment]
@UTKARSHS1994 hi Utkarsh,

You are on the right track with using the `boxes=False` argument in YOLOv8 for segmentation inference, which indeed should remove the bounding boxes and return only the masks.

From the code snippet you've provided, it looks correct on the surface. However, there might be various reasons why the bounding boxes are still appearing. One possible reason could be related to your specific model training, where the bounding boxes could be inherently part of the segmentation mask. 

Moreover, ensure that when you are performing prediction, you are indeed using the trained model from the same segmentation task, and not a different model trained to do detection. 

Remember, it's important to match the correct model with the corresponding task in order to produce the expected results. 

Don't hesitate to reach out if you have more questions or if there's anything else we can do to help!

Best Regards,
Glenn

**Comment by UTKARSHS1994** [comment]
@glenn-jocher Thanks Glenn for quick reply.

Keeping in mind the suggestions you made, I retrained the model again as following (check the arguments underlined in red):

![image](https://github.com/ultralytics/ultralytics/assets/54258787/34290aa8-1030-47c4-a895-0ac6e2fb7065)

After training the model for 100 epochs, I did inferencing on test images as follows:

![image](https://github.com/ultralytics/ultralytics/assets/54258787/ed6e76ef-4de7-4224-97f0-705f3a4b9581)

And got the following results :

![image](https://github.com/ultralytics/ultralytics/assets/54258787/ac3abe2c-c4a1-46f5-a385-3d2b8be5ab29)

I am not sure why boxes are still visible and what can be done to get rid of them.

Regards,
Utkarsh

**Comment by glenn-jocher** [comment]
@UTKARSHS1994 hi Utkarsh,

As per your description, you've followed the steps correctly and yet the boxes are still appearing in your results, which is indeed unexpected with the 'boxes=False' parameter. This may be due to reasons related to the refinement of the segmentation task or peculiarities with the custom dataset used.

However, without any further errors or additional information, it's challenging for me to pinpoint the exact issue. It's worth noting that it's crucial to match the correct model with the corresponding task (e.g., using a model trained for segmentation for segmentation inference) in order to produce the expected results.

You might want to ensure the dataset, and the annotations are in the right format and the model checkpoint being used is the correct one. Also, it's always a good practice to update to the latest version of YOLOv8 as there are ongoing improvements and bug fixes which might solve your specific issue.

I know it's frustrating when you've done everything right and yet the output doesn't match the expectation. We appreciate your patience, and the YOLOv8 team is committed to assist you further to resolve this issue.

Best Regards,
Glenn

**Comment by UTKARSHS1994** [comment]
@glenn-jocher Hi Glenn,

I tried to get inference on one of the images using available YOLOv8 model as follows. I still get the boxes:

![image](https://github.com/ultralytics/ultralytics/assets/54258787/9e93ade3-e900-4382-a290-a46d2e7096f8)

Got the following:

![image](https://github.com/ultralytics/ultralytics/assets/54258787/7ad8aef7-d62b-43eb-aaf4-001c161d59eb)

 I tried the above both in Kaggle and Colab. How can I use a pretrained model to perform segmentation without bounding boxes? I am sharing this test image for your reference below.

Regards,
Utkarsh

![d_r_122_](https://github.com/ultralytics/ultralytics/assets/54258787/f3adce7d-27ac-4958-8def-c4303cfdc43c)

**Comment by glenn-jocher** [comment]
@UTKARSHS1994 hi Utkarsh,

Thank you for reaching out and for using the Ultralytics YOLOv8. It's strange that you are still seeing boxes in the output even after setting `boxes=False` during inference. The issue you're facing is unexpected since setting `boxes=False` should indeed remove the bounding boxes and return only the masks in the output.

There could be various reasons why the bounding boxes are still appearing. Without seeing any error messages it's hard to know for sure, but possible reasons could include version differences in the underlying libraries or incorrect model configurations.

While I understand your problem and can appreciate how frustrating it must be, it is difficult for me to reproduce or diagnose without the exact code and environment details you're using on Kaggle and Colab.

Please ensure you're using the latest version of YOLOv8. I will inform the team to look into this issue and run more tests to avoid such situations in the future. Appreciate your patience on this.

Best regards,
Glenn

**Comment by UTKARSHS1994** [comment]
@glenn-jocher Hi Glenn,

I appreciate that you are diligently following up with me. I wish to thank you for this. 

I tried inferencing right now with the latest version of ultralytics 8.0.180  
The code needed to reproduce it is below. I have already provided the test_image in my previous response. You can download the image and just change the path of the  image in the variable test_image:

#### Code starts here
import matplotlib.pyplot as plt
import cv2
!pip install ultralytics
from ultralytics import YOLO

model=YOLO('yolov8m-seg.pt')
plt.figure(figsize=(30,30))
ax=plt.subplot(1,2,1)
test_image='/kaggle/input/semantic-segmentation-of-underwater-imagery-suim/TEST/images/d_r_122_.jpg'

plt.imshow(cv2.cvtColor(cv2.imread(test_image), cv2.COLOR_BGR2RGB)) 
plt.title("Actual image", fontsize = 40)

res = model(test_image, boxes=False)
res_plotted = res[0].plot()
ax=plt.subplot(1,2,2)

plt.imshow(res_plotted)
plt.title("Image with predictions", fontsize = 40)
#### Code ends here


Hope now the problem is reproducible.

Regards,
Utkarsh

**Comment by glenn-jocher** [comment]
@UTKARSHS1994 Hi Utkarsh,

Thank you for providing the code and further context to your issue. This will certainly help in our attempt to reproduce and understand what could be going wrong. 

As you've written, on face value, setting the parameter `boxes = False` should indeed present the output without any bounding boxes. It's quite unusual that the bounding boxes are still visible in your output, particularly that you're correctly using the segmentation model with the YOLOv8 framework. 

At this point, without any error messages and given the code you've provided, it's challenging to know exactly why this is occurring. Several factors could be in play, such as the specifics of the training set used, the variant of YOLOv8 model used, or some unexpected behavior in the prediction. 

However, please rest assured that your query has been noted and we will further investigate it. We advise you to keep your software versions up to date in the meanwhile.

Thank you once again for bringing this to our attention and for providing more detailed context; this will undoubtedly aid in the investigation. 

Best Regards,
Glenn

**Comment by tsampazk** [comment]
Hey @UTKARSHS1994. I ran into the same issue basically by misunderstanding the usage of the arguments in `predict`. Looking at your code, you are setting `boxes=False` on the prediction step, which is valid when you are also passing `show=True`, which causes the prediction method to show/visualize the image with the results, without the bboxes. 

In your code snippets you are using `results.plot()` to get the annotated image. You should use `results[0].plot(boxes=False)` to force plot to not annotate the image with the bounding boxes. Of course, all this refers to the visualization, you can access the masks and/or bounding boxes from the `results` object in code.

If you want to more easily just visualize the results straight from the prediction, use `results=model('path_to_image', boxes=False, show=True)

Hope this helps.

**Comment by glenn-jocher** [comment]
Hello @tsampazk,

I see you're trying to get rid of the bounding boxes while plotting the outputs of a YOLOv8 segmentation model prediction. Apologies for the confusion caused, the `boxes=False` parameter in the `predict()` function only influences the display when you opt for real-time rendering of the outputs using the `show=True` keyword argument.

However, when you're using `plot()` function to visualize the annotated image, you'll have to direct the function explicitly not to annotate the image with bounding boxes. You can do this by calling `results[0].plot(boxes=False)` instead.

In case you prefer visualizing the results directly from the model's predictions, you can also use `results = model('path_to_image', boxes=False, show=True)`. This function call would directly display the output image without any bounding boxes.

Your understanding is crucial to us, and we hope this clarification helps you with your project. Please feel free to reach out if you have more questions or need further clarification.

Best regards,
Glenn

**Comment by tsampazk** [comment]
Hey @glenn-jocher, i have figured this out tinkering with the code and the docs. From your answer i can tell that my understanding is correct and Utkarsh's issue can be solved this way. Thanks anyway!

**Comment by glenn-jocher** [comment]
Hello @tsampazk,

I'm glad to hear that you managed to figure things out and improve your understanding through exploring the code and documentation. Your approach could indeed solve Utkarsh's issue.

Your initiative and proactive problem-solving are much appreciated. I'm also pleased that our initial communication could provide some guidance. I am happy to assist if you face more questions or issues regarding YOLOv8 in the future.

Best Regards,
Glenn

**Comment by UTKARSHS1994** [comment]
Hey @tsampazk , Thanks for your valuable solution to this problem. Visualizing the results as `results[0].plot(boxes=False`) is indeed useful as it shows the results without bounding boxes.
However, it also removes the labels as well. I expect the results to show masks with labels but without bounding box. I also tried `results[0].plot(boxes=False, labels=True)` but the labels do not appear. It seems like labels are available only with bounding-boxes. Let me know if we can do anything for this.

Thanks again.

Regards,
Utkarsh

**Comment by glenn-jocher** [comment]
Hello @UTKARSHS1994,

Thank you for your feedback about this. At the moment, in YOLOv8, the labels are indeed tied with the bounding boxes. This means that if the bounding boxes are removed, the labels are also removed. This is because the labels are positioned at the top-left corner of the bounding boxes by default, and since there is no bounding box, the labels find no placement.

As of now, there's no built-in option in the plot function to display labels independent of bounding boxes. This behavior is embedded into the architecture of the YOLOv8 model and it's non-trivial to change this without major modifications to the underlying structure.

However, your feedback gives us insight into a potential feature that could enhance the user experience. We will consider this for our future updates.

Thanks again for bringing this to our attention.

Best Regards,
Glenn

**Comment by UTKARSHS1994** [comment]
Ah, I see! I am glad that this might be considered in a future update. I will surely be looking forward to it. This will surely improve user experience.

Thanks @glenn-jocher for your time and continuous involvement to address these issues. I really appreciate that.

Regards,
Utkarsh

**Comment by glenn-jocher** [comment]
Hello @UTKARSHS1994,

I'm really pleased to hear your understanding and enthusiasm for potential updates to the YOLOv8 model. It's feedback and constructive dialogues like these that continue to make our work better.

Thank you for your kind words and recognition of our efforts. We're grateful for users like you who actively contribute to improving our repository. Your participation is what helps us push the boundaries and continually enhance our models.

I look forward to more insightful interactions in the future.

Best regards, 
Glenn
